docs oidc cookie encryption (#8112)
solo-io/gloo: docs/content/guides/security/auth/extauth/oauth/_index.md

If the authorization server has a service error, Gloo logs out the user, but does not retry revoking the access token. Check the logs and your identity provider for errors, and manually revoke the access token.
{{% /notice %}}

## Sessions in Cookies

> You can store the id token, access token in cookies on the client side. To do this you will need to configure the `session` with the `cookie` configurations. Here, we set the `keyPrefix` to add a prefix to the cookie name.

Reviewer: I think it's worth it calling out here, or somewhere what cookie based session limitations exist. I thought a core issue we've run into recently is where contents of those tokens exceed the 4k limit for a cookie

> In addition, you can configure the settings to encrypt the cookie values as well, using the new `cipherConfig`. This is available in Gloo v1.15.0 and later. You can generate a key using the following command `glooctl create secret encryptionkey --name my-encryption-key --key "an example of an encryption key1"`. Then you can set the `keyRef` for the generated `encryptionkey`. NOTE: that the key has to be 32 bytes in length for it to work. The encryption only works on cookie sessions, this is not applied to redis.

Reviewer: It may be valuable to explain and provide an example of setting up cookie auth without encryption. Then we can outline what security risks that poses and why you would want to use encrption. And then we add this example of how to enable it

Example configuration:
{{< highlight yaml "hl_lines=19-25" >}}
apiVersion: enterprise.gloo.solo.io/v1
kind: AuthConfig
metadata:
  name: oidc-dex
  namespace: gloo-system
spec:
  configs:
  - oauth2:
      oidcAuthorizationCode:
        appUrl: http://localhost:8080/
        callbackPath: /callback
        clientId: gloo
        clientSecretRef:
          name: oauth
          namespace: gloo-system
        issuerUrl: http://dex.gloo-system.svc.cluster.local:32000/
        scopes:
        - email
        session:
          cipherConfig:
            keyRef:
              name: my-encryption-key
              namespace: gloo-system
          cookie:
            keyPrefix: "my_cookie_prefix"
{{< /highlight >}}

## Sessions in Redis

By default, the tokens will be saved in a secure client side cookie.
Gloo can instead use Redis to save the OIDC tokens, and set a randomly generated session id in the user's cookie.

> Going forward, we will be using examples of OIDC using a redis session.

Reviewer: Is this because redis is our recommended solution? Or just that it's easier to document configuration with a certain persistence layer and not worry about both

Author: the examples in the documentation all use redis.

If the authorization server has a service error, Gloo logs out the user, but does not retry revoking the access token. Check the logs and your identity provider for errors, and manually revoke the access token.
{{% /notice %}}

## Sessions in Cookies

You can store the id token, access token, and other cookies on the client side. To do this you will need to configure the `session` with the `cookie` configurations. Here, we set the `keyPrefix` to add a prefix to the cookie name. This will expose the raw cookie value to the client, for more secure options please look at the next steps in symmetric cookie encryption.

> Current limitations apply to the size of cookies. Cookies cannot be larger than 4KB. Approaches to limit the size of the cookies or using redis sessions are solutions around this.

Reviewer 2: Can we add this to the production guidelines doc? Namely that the 4k limit exists and if your cookies grow we need to split them or move to non-cookie based setups.

{{% /notice %}}

Example configuration:
{{< highlight yaml "hl_lines=19-21" >}}
apiVersion: enterprise.gloo.solo.io/v1
kind: AuthConfig
metadata:
  name: oidc-dex
  namespace: gloo-system
spec:
  configs:
  - oauth2:
      oidcAuthorizationCode:
        appUrl: http://localhost:8080/
        callbackPath: /callback
        clientId: gloo
        clientSecretRef:
          name: oauth
          namespace: gloo-system
        issuerUrl: http://dex.gloo-system.svc.cluster.local:32000/
        scopes:
        - email
        session:
          cookie:
            keyPrefix: "my_cookie_prefix"
{{< /highlight >}}

### Symmetric cookie encryption

By default, the tokens that are sent in the cookie header are not encrypted and can be manipulated through malicious attacks. To encrypt the cookie values, you can add a `cipherConfig` section to your session configuration as shown in the following example. 

{{% notice note %}}
Setting the `cipherConfig` attribute is supported in Gloo Edge version 1.15 and later and can be used only to encrypt cookie sessions. You cannot use this feature to encrypt Redis sessions. 
{{% /notice %}}

1. Create a secret with your encryption key. 
   ```shell

>    glooctl create secret encryptionkey --name my-encryption-key --key "an example of an encryption key"

Author: `"an example of an encryption key"` is 31 bytes long, must be 32 bytes long.
```suggestion
   glooctl create secret encryptionkey --name my-encryption-key --key "an example of an encryption key1"
```

   ```

> 2. Reference the secret in the `cipherConfig` section of your authconfig. Note that they key must be 32 bytes in length. 

Reviewer 3: ```suggestion
2. Reference the secret in the `cipherConfig` section of your authconfig. Note that the key must be 32 bytes in length. 
```

> 2. Reference the secret in the `cipherConfig` section of your authconfig. Note that they key must be 32 bytes in length. 

Author: ```suggestion
1. Create a secret with your encryption key.  Note that they key must be 32 bytes in length.
   ```shell
   glooctl create secret encryptionkey --name my-encryption-key --key "an example of an encryption key"
   ```

2. Reference the secret in the `cipherConfig` section of your authconfig.
```

>    {{< highlight yaml "hl_lines=2-5" >}}
>    session:
>      cipherConfig:
>        keyRef:
>          name: my-encryption-key
>          namespace: gloo-system
>      cookie:
>        keyPrefix: "my_cookie_prefix"
>    {{< /highlight >}}

Reviewer 3: Could the YAML indentation show the full path of where this would go?
```suggestion
   {{< highlight yaml "hl_lines=8-11" >}}
   ...
   kind: AuthConfig
   spec:
     configs:
     - oauth2:
          oidcAuthorizationCode:
            session:
              cipherConfig:
                keyRef:
                  name: my-encryption-key
                  namespace: gloo-system
                cookie:
                  keyPrefix: "my_cookie_prefix"
   {{< /highlight >}}
```
